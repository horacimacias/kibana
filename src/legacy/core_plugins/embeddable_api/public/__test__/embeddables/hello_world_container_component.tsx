/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import { Container, Embeddable } from 'plugins/embeddable_api/index';
import React, { Component, RefObject } from 'react';

import { EuiFlexGroup, EuiFlexItem } from '@elastic/eui';
import { EuiSpacer } from '@elastic/eui';

interface Props {
  container: Container;
}

interface State {
  embeddables: { [key: string]: Embeddable };
  loaded: { [key: string]: boolean };
}

export class HelloWorldContainerComponent extends Component<Props, State> {
  private roots: { [key: string]: RefObject<HTMLDivElement> } = {};
  private mounted: boolean = false;
  private inputSubscription?: Subscription;
  private outputSubscription?: Subscription;

  public constructor(props: Props) {
    super(props);

    Object.values(this.props.container.getInput().panels).forEach(panelState => {
      this.roots[panelState.embeddableId] = React.createRef();
    });
    this.state = {
      loaded: {},
      embeddables: {},
    };
  }

  public async componentDidMount() {
    this.mounted = true;

    Object.values(this.props.container.getInput().panels).forEach(panelState => {
      this.renderEmbeddable(panelState.embeddableId);
    });

    this.inputSubscription = this.props.container.getInput$().subscribe(() => {
      Object.values(this.props.container.getInput().panels).forEach(async panelState => {
        if (this.roots[panelState.embeddableId] === undefined) {
          this.roots[panelState.embeddableId] = React.createRef();
        }

        if (this.state.embeddables[panelState.embeddableId] === undefined) {
          const embeddable = await this.props.container.getEmbeddable(panelState.embeddableId);
          const node = this.roots[panelState.embeddableId].current;
          if (this.mounted && node !== null && embeddable) {
            embeddable.renderInPanel(node, this.props.container);

            this.setState(prevState => ({
              loaded: { ...prevState.loaded, [panelState.embeddableId]: true },
            }));
          }
        }
      });
    });

    this.outputSubscription = this.props.container.getOutput$().subscribe(() => {
      const embeddablesLoaded = this.props.container.getOutput().embeddableLoaded;
      Object.keys(embeddablesLoaded).forEach(async id => {
        const loaded = embeddablesLoaded[id];
        if (loaded && !this.state.loaded[id]) {
          const embeddable = await this.props.container.getEmbeddable(id);
          const node = this.roots[id].current;
          if (this.mounted && node !== null && embeddable) {
            embeddable.renderInPanel(node, this.props.container);
            this.setState({ loaded: embeddablesLoaded });
          }
        }
      });
    });
  }

  public componentWillUnmount() {
    this.props.container.destroy();

    if (this.inputSubscription) {
      this.inputSubscription.unsubscribe();
    }
    if (this.outputSubscription) {
      this.outputSubscription.unsubscribe();
    }
  }

  public renderList() {
    const list = Object.values(this.props.container.getInput().panels).map(panelState => {
      const item = (
        <EuiFlexItem key={panelState.embeddableId}>
          <div style={{ height: '400px' }} ref={this.roots[panelState.embeddableId]} />
        </EuiFlexItem>
      );
      return item;
    });
    return list;
  }

  public render() {
    return (
      <div>
        <h2>HELLO WORLD! These are my precious embeddable children:</h2>
        <EuiSpacer size="l" />
        <EuiFlexGroup>{this.renderList()}</EuiFlexGroup>
      </div>
    );
  }

  private async renderEmbeddable(id: string) {
    if (this.state.embeddables[id] !== undefined) {
      return;
    }

    if (this.roots[id] === undefined) {
      this.roots[id] = React.createRef();
    }

    if (this.state.embeddables[id] === undefined) {
      const embeddable = await this.props.container.getEmbeddable(id);
      const node = this.roots[id].current;
      if (this.mounted && node !== null && embeddable) {
        embeddable.renderInPanel(node, this.props.container);

        this.setState(prevState => ({
          loaded: { ...prevState.loaded, [id]: true },
          embeddables: {
            [id]: embeddable,
          },
        }));
      }
    }
  }
}
