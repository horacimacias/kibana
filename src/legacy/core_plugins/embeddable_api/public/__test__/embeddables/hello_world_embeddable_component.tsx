/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import React from 'react';
import { HelloWorldEmbeddable } from './hello_world_embeddable';

interface Props {
  helloWorldEmbeddable: HelloWorldEmbeddable;
}

interface State {
  fullName: string;
}

export class HelloWorldEmbeddableComponent extends React.Component<Props, State> {
  private unsubscribe?: () => void;
  private mounted: boolean = false;

  constructor(props: Props) {
    super(props);
    this.state = {
      fullName: '',
    };
  }

  componentDidMount() {
    this.mounted = true;
    this.unsubscribe = this.props.helloWorldEmbeddable.subscribeToOutputChanges(() => {
      if (this.mounted) {
        this.setState({ fullName: this.props.helloWorldEmbeddable.getOutput().name });
      }
    });
  }

  componentWillUnmount() {
    if (this.unsubscribe) {
      this.unsubscribe();
    }
    this.mounted = false;
  }

  render() {
    return (
      <div>
        Hello World!
        <br /> Sincerly, ${this.state.fullName}
      </div>
    );
  }
}
