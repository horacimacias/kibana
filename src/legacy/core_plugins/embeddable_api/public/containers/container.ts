/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import uuid from 'uuid';
import {
  Embeddable,
  EmbeddableFactoryRegistry,
  EmbeddableInput,
  EmbeddableOutput,
  ErrorEmbeddable,
  EmbeddableFactory,
} from '../embeddables';
import { ViewMode } from '../types';

export interface PanelState<E extends { [key: string]: unknown } = { [key: string]: unknown }> {
  savedObjectId?: string;

  embeddableId: string;
  // The type of embeddable in this panel. Will be used to find the factory in which to
  // load the embeddable.
  type: string;

  // Stores input for this embeddable that is specific to this embeddable. Other parts of embeddable input
  // will be derived from the container's input. **Any state in here will override any state derived from
  // the container.**
  partialInput: E; // TODO: rename explicitInput vs derivedInput that is taken from the container state.
}

export interface ContainerOutput extends EmbeddableOutput {
  embeddableLoaded: { [key: string]: boolean };
}

type NonNeverPropertyNames<T> = { [K in keyof T]: T[K] extends never ? never : K }[keyof T];
type ObjectWithoutNever<T> = Pick<T, NonNeverPropertyNames<T>>;

export type EmbeddableInputMissingFromContainer<
  EI extends EmbeddableInput,
  CEI extends { [key: string]: unknown }
> = ObjectWithoutNever<{ [key in keyof EI]: CEI[key] extends EI[key] ? never : EI[key] }>;

export interface ContainerInput extends EmbeddableInput {
  hidePanelTitles?: boolean;
  panels: {
    [key: string]: PanelState;
  };
}

export abstract class Container<
  CEI extends Partial<EmbeddableInput> = {},
  EO extends EmbeddableOutput = EmbeddableOutput,
  I extends ContainerInput = ContainerInput,
  O extends ContainerOutput = ContainerOutput
> extends Embeddable<I, O> {
  public readonly isContainer: boolean = true;
  protected readonly embeddables: {
    [key: string]: Embeddable<EmbeddableInput, EO> | ErrorEmbeddable;
  } = {};
  public readonly embeddableFactories: EmbeddableFactoryRegistry;

  constructor(
    type: string,
    input: I,
    output: O,
    embeddableFactories: EmbeddableFactoryRegistry,
    parent?: Container
  ) {
    super(type, input, output, parent);
    this.embeddableFactories = embeddableFactories;
    this.initializeEmbeddables();
  }

  public getPartialInput(embeddableId: string) {
    return this.input.panels[embeddableId].partialInput;
  }

  public getViewMode() {
    return this.input.viewMode ? this.input.viewMode : ViewMode.EDIT;
  }

  public getHidePanelTitles() {
    return this.input.hidePanelTitles ? this.input.hidePanelTitles : false;
  }

  public updateInput(changes: Partial<I>): void {
    super.updateInput(changes);
  }

  public updateEmbeddableInput<EEI extends EmbeddableInput = EmbeddableInput>(
    id: string,
    changes: Partial<EEI>
  ) {
    if (!this.input.panels[id]) {
      throw new Error();
    }
    const panels = {
      panels: {
        ...this.input.panels,
        [id]: {
          ...this.input.panels[id],
          partialInput: {
            ...this.input.panels[id].partialInput,
            ...changes,
          },
        },
      },
    };
    this.updateInput(panels as Partial<I>);
  }

  public async addNewEmbeddable<
    EEI extends EmbeddableInput = EmbeddableInput,
    E extends Embeddable<EEI, EO> = Embeddable<EEI, EO>
  >(type: string, partialInput: Partial<EEI>): Promise<E | ErrorEmbeddable> {
    const factory = this.embeddableFactories.getFactoryByName(type) as EmbeddableFactory<
      EEI,
      EO,
      E
    >;
    const panelState = this.createNewPanelState<EEI>(factory, partialInput);
    this.setPanelState<EEI>(panelState);
    const embeddable = await factory.create<Container<CEI, EO, I, O>>(
      this.getInputForEmbeddable<EEI>(panelState.embeddableId),
      this
    );
    this.embeddables[embeddable.id] = embeddable;
    this.updateOutput({
      ...this.output,
      embeddableLoaded: {
        [panelState.embeddableId]: true,
      },
    });
    return embeddable;
  }

  private createNewExplicitEmbeddableInput<
    EEI extends EmbeddableInput = EmbeddableInput,
    E extends Embeddable<EEI, EO> = Embeddable<EEI, EO>
  >(id: string, factory: EmbeddableFactory<EEI, EO, E>, partial: Partial<EEI> = {}) {
    const inheritedInput = this.getInheritedInput(id) as { [key: string]: unknown };
    const defaults = factory.getDefaultInputParameters() as { [key: string]: unknown };

    // Container input overrides defaults.
    const explicitInput: { [key: string]: unknown } = partial;
    Object.keys(defaults).forEach(key => {
      if (inheritedInput[key] === undefined && explicitInput[key] === undefined) {
        explicitInput[key] = defaults[key];
      }
    });
    return explicitInput;
  }

  public async addSavedObjectEmbeddable<
    EEI extends EmbeddableInput = EmbeddableInput,
    E extends Embeddable<EEI, EO> = Embeddable<EEI, EO>
  >(type: string, savedObjectId: string): Promise<E | ErrorEmbeddable> {
    const factory = this.embeddableFactories.getFactoryByName(type) as EmbeddableFactory<
      EEI,
      EO,
      E
    >;
    const panelState = this.createNewPanelState<EEI>(factory);
    panelState.savedObjectId = savedObjectId;
    this.setPanelState<EEI>(panelState);
    const embeddable = await factory.createFromSavedObject(
      savedObjectId,
      this.getInputForEmbeddable<EEI>(panelState.embeddableId),
      this
    );
    this.embeddables[embeddable.id] = embeddable;

    this.updateOutput({
      ...this.output,
      embeddableLoaded: {
        [panelState.embeddableId]: true,
      },
    });
    return embeddable;
  }

  public removeEmbeddable(embeddableId: string) {
    const embeddable = this.getEmbeddable(embeddableId);
    embeddable.destroy();
    delete this.embeddables[embeddableId];

    const changedInput: { panels: { [key: string]: PanelState } } = {
      panels: {},
    };
    Object.values(this.input.panels).forEach(panel => {
      if (panel.embeddableId !== embeddableId) {
        changedInput.panels[panel.embeddableId] = panel;
      }
    });
    this.updateInput({ ...changedInput } as Partial<I>);
  }

  public getEmbeddable<E extends Embeddable<EmbeddableInput, EO> = Embeddable<EmbeddableInput, EO>>(
    id: string
  ): E {
    return this.embeddables[id] as E;
  }

  private setPanelState<EEI extends EmbeddableInput = EmbeddableInput>(
    panelState: PanelState<EmbeddableInputMissingFromContainer<EEI, CEI>>
  ) {
    this.updateInput({
      panels: {
        ...this.input.panels,
        [panelState.embeddableId]: panelState,
      },
    } as Partial<I>);
  }

  private updatePanelState<EEI extends EmbeddableInput = EmbeddableInput>(
    panelState: PanelState<EmbeddableInputMissingFromContainer<EEI, CEI>>
  ) {
    this.updateInput({
      panels: {
        ...this.input.panels,
        [panelState.embeddableId]: {
          ...this.input.panels[panelState.embeddableId],
          ...panelState,
        },
      },
    } as Partial<I>);
  }

  protected async loadEmbeddable<EEI extends EmbeddableInput = EmbeddableInput>(
    panelState: PanelState<EmbeddableInputMissingFromContainer<EEI, CEI>>
  ) {
    if (this.input.panels[panelState.embeddableId] === undefined) {
      throw new Error(`Panel with id ${panelState.embeddableId} does not exist in this container`);
    }

    const factory = this.embeddableFactories.getFactoryByName<EmbeddableFactory<EEI, EO>>(
      panelState.type
    );

    const embeddable = panelState.savedObjectId
      ? await factory.createFromSavedObject(
          panelState.savedObjectId,
          this.getInputForEmbeddable<EEI>(panelState.embeddableId),
          this
        )
      : await factory.create(this.getInputForEmbeddable<EEI>(panelState.embeddableId), this);
    this.embeddables[embeddable.id] = embeddable;
    this.updatePanelState<EEI>(panelState);

    this.updateOutput({
      embeddableLoaded: {
        [panelState.embeddableId]: true,
      },
    } as Partial<O>);
  }

  protected createNewPanelState<EEI extends EmbeddableInput = EmbeddableInput>(
    factory: EmbeddableFactory<EEI, EO>,
    partial: Partial<EEI> = {}
  ): PanelState<EmbeddableInputMissingFromContainer<EEI, CEI>> {
    const embeddableId = partial.id || uuid.v4();

    const explicitInput = this.createNewExplicitEmbeddableInput<EEI>(
      embeddableId,
      factory,
      partial
    ) as EmbeddableInputMissingFromContainer<EEI, CEI>;

    return {
      type: factory.name,
      embeddableId,
      partialInput: {
        id: embeddableId,
        ...explicitInput,
      },
    };
  }

  protected getPanelState<EI>(embeddableId: string) {
    if (this.input.panels[embeddableId] === undefined) {
      throw new Error(`No embeddable with id ${embeddableId}, this  ${JSON.stringify(this.input)}`);
    }
    const panelState: PanelState = this.input.panels[embeddableId];
    return panelState as PanelState<EI>;
  }

  protected abstract getInheritedInput(id: string): CEI;

  public getExplicitEmbeddableInput(id: string) {
    return this.input.panels[id].partialInput;
  }

  public getInputForEmbeddable<EEI extends EmbeddableInput = EmbeddableInput>(
    embeddableId: string
  ): EEI {
    const containerInput: CEI = this.getInheritedInput(embeddableId);
    const panelState = this.getPanelState<EmbeddableInputMissingFromContainer<EEI, CEI>>(
      embeddableId
    );

    const partialInput = panelState.partialInput as { [key: string]: unknown };
    const partialInputWithNoUndefinedParameters: { [key: string]: unknown } = {};

    Object.keys(panelState.partialInput).forEach(key => {
      if (partialInput[key] !== undefined) {
        partialInputWithNoUndefinedParameters[key] = partialInput[key];
      }
    });

    return ({
      ...containerInput,
      ...partialInputWithNoUndefinedParameters,
      // Typescript has difficulties with inferring this type but it is accurate with all
      // tests I tried. Could probably be revisted with future releases of TS to see if
      // it can accurately infer the type.
    } as unknown) as EEI;
  }

  private async initializeEmbeddables() {
    const promises = Object.values(this.input.panels).map(async panel => {
      await this.loadEmbeddable<EmbeddableInput>(panel as PanelState<
        EmbeddableInputMissingFromContainer<EmbeddableInput, CEI>
      >);
    });
    await Promise.all(promises);
  }
}
