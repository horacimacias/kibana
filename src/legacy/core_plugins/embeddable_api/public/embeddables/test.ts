/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

/* eslint-disable max-classes-per-file */
import * as Rx from 'rxjs';

interface AnimalData {
  sound: string;
}

// class Animal<D extends AnimalData = AnimalData> {
//   public readonly data: D;

//   constructor(data: D) {
//     this.data = data;
//   }

//   debug() {
//     debug(this);
//   }
// }

// function debug(animal: Animal) {
//   console.log(animal.data);
// }

class Animal<D extends {} = {}> {
  public readonly data: D;
  private readonly listeners: Array<(d: D) => void> = [];

  constructor(data: D) {
    this.data = data;
  }

  debug() {
    debug(this);
  }
}

function debug(animal: Animal) {
  console.log(animal.data);
}

// class Cat extends Animal<{ sound: string }> {}

// doSomething(new Cat({ sound: 'meow' }));
