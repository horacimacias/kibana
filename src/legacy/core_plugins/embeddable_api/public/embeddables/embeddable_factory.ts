/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { SavedObjectMetaData } from 'ui/saved_objects/components/saved_object_finder';
import { SavedObjectAttributes } from '../../../../server/saved_objects';
import { Embeddable, EmbeddableInput, EmbeddableOutput } from './embeddable';
import { ErrorEmbeddable } from './error_embeddable';
import { Container } from '../containers';

export interface EmbeddableInstanceConfiguration {
  id: string;
  savedObjectId?: string;
}

export type AnyEmbeddableFactory = EmbeddableFactory;

export interface PropertySpec {
  displayName: string;
  accessPath: string;
  id: string;
  description: string;
  value?: string;
}

export interface OutputSpec {
  [key: string]: PropertySpec;
}

/**
 * The EmbeddableFactory creates and initializes an embeddable instance
 */
export abstract class EmbeddableFactory<
  I extends EmbeddableInput = EmbeddableInput,
  O extends EmbeddableOutput = EmbeddableOutput,
  E extends Embeddable<I, O> = Embeddable<I, O>,
  T extends SavedObjectAttributes = SavedObjectAttributes
> {
  public readonly name: string;
  public readonly savedObjectMetaData?: SavedObjectMetaData<T>;

  /**
   *
   * @param name - a unique identified for this factory, which will be used to map an embeddable spec to
   * a factory that can generate an instance of it.
   */
  constructor({
    name,
    savedObjectMetaData,
  }: {
    name: string;
    savedObjectMetaData?: SavedObjectMetaData<T>;
  }) {
    this.name = name;
    this.savedObjectMetaData = savedObjectMetaData;
  }

  public abstract getOutputSpec(): OutputSpec;

  public isEditable() {
    return true;
  }

  public getDefaultInputParameters(): Partial<I> {
    return {};
  }

  /**
   */
  public createFromSavedObject(
    savedObjectId: string,
    partialInput: Partial<I>,
    parent?: Container
  ): Promise<E | ErrorEmbeddable> {
    throw new Error(`Creation from saved object not supported by type ${this.name}`);
  }

  /**
   */
  public abstract create<C extends Container = Container>(
    initialInput: I,
    parent?: C
  ): Promise<E | ErrorEmbeddable>;
}
