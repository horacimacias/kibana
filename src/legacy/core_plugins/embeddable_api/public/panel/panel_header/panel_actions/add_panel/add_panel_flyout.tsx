/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import { i18n } from '@kbn/i18n';
import { FormattedMessage } from '@kbn/i18n/react';
import React from 'react';
import { toastNotifications } from 'ui/notify';
import {
  SavedObjectFinder,
  SavedObjectMetaData,
} from 'ui/saved_objects/components/saved_object_finder';

import {
  EuiFlexGroup,
  EuiFlexItem,
  EuiFlyout,
  EuiFlyoutBody,
  EuiFlyoutFooter,
  EuiFlyoutHeader,
  // @ts-ignore
  EuiSuperSelect,
  EuiTitle,
  EuiText,
} from '@elastic/eui';

import { SavedObjectAttributes } from 'src/legacy/server/saved_objects';
import { Container } from '../../../../containers';

interface Props {
  onClose: () => void;
  container: Container;
}

export class AddPanelFlyout extends React.Component<Props> {
  private lastToast: any;

  constructor(props: Props) {
    super(props);
  }

  public showToast = (name: string) => {
    // To avoid the clutter of having toast messages cover flyout
    // close previous toast message before creating a new one
    if (this.lastToast) {
      toastNotifications.remove(this.lastToast);
    }

    this.lastToast = toastNotifications.addSuccess({
      title: i18n.translate(
        'kbn.embeddables.addPanel.savedObjectAddedToContainerSuccessMessageTitle',
        {
          defaultMessage: '{savedObjectName} was added',
          values: {
            savedObjectName: name,
          },
        }
      ),
      'data-test-subj': 'addObjectToContainerSuccess',
    });
  };

  public createNewEmbeddable = async (type: string) => {
    this.props.onClose();
    const embeddable = await this.props.container.addNewEmbeddable(type, {});
    this.showToast(embeddable.getInput().title || '');

    // This redirect is to preserve existing visualization editing flow. In the future, we would
    // ideally jump into edit mode while staying in the current location in the container.
    const editUrl = embeddable.getOutput().editUrl;
    if (editUrl) {
      location.assign(editUrl);
    }
  };

  public onAddPanel = async (id: string, type: string, name: string) => {
    this.props.container.addSavedObjectEmbeddable(type, id);

    this.showToast(name);
  };

  private getSelectCreateNewOptions() {
    return [
      {
        value: 'createNew',
        inputDisplay: (
          <EuiText>
            <FormattedMessage
              id="kbn.embeddables.addPanel.createNewDefaultOption"
              defaultMessage="Create new ..."
            />
          </EuiText>
        ),
      },

      ...Object.values(this.props.container.embeddableFactories.getFactories())
        .filter(factory => factory.isEditable())
        .map(factory => ({
          inputDisplay: (
            <EuiText>
              <FormattedMessage
                id="kbn.embeddables.addPanel.createNew"
                defaultMessage="Create new {factoryName}"
                values={{
                  factoryName: factory.name,
                }}
              />
            </EuiText>
          ),
          value: factory.name,
          'data-test-subj': `createNew-${factory.name}`,
        })),
    ];
  }

  public render() {
    return (
      <EuiFlyout ownFocus onClose={this.props.onClose} data-test-subj="dashboardAddPanel">
        <EuiFlyoutHeader hasBorder>
          <EuiTitle size="m">
            <h2>
              <FormattedMessage id="kbn.embeddables.addPanelsTitle" defaultMessage="Add panels" />
            </h2>
          </EuiTitle>
        </EuiFlyoutHeader>
        <EuiFlyoutBody>
          <SavedObjectFinder
            onChoose={this.onAddPanel}
            savedObjectMetaData={
              Object.values(this.props.container.embeddableFactories.getFactories())
                .filter(embeddableFactory => Boolean(embeddableFactory.savedObjectMetaData))
                .map(({ savedObjectMetaData }) => savedObjectMetaData) as Array<
                SavedObjectMetaData<SavedObjectAttributes>
              >
            }
            showFilter={true}
            noItemsMessage={i18n.translate('kbn.embeddables.addPanel.noMatchingObjectsMessage', {
              defaultMessage: 'No matching objects found.',
            })}
          />
        </EuiFlyoutBody>
        <EuiFlyoutFooter>
          <EuiFlexGroup justifyContent="flexEnd">
            <EuiFlexItem grow={true}>
              <EuiSuperSelect
                data-test-subj="createNew"
                options={this.getSelectCreateNewOptions()}
                valueOfSelected="createNew"
                onChange={(value: string) => this.createNewEmbeddable(value)}
              />
            </EuiFlexItem>
          </EuiFlexGroup>
        </EuiFlyoutFooter>
      </EuiFlyout>
    );
  }
}
