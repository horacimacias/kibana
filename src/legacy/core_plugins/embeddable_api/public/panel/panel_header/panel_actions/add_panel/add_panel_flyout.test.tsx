/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

jest.mock('ui/metadata', () => ({
  metadata: {
    branch: 'my-metadata-branch',
    version: 'my-metadata-version',
  },
}));

jest.mock('ui/capabilities', () => ({
  uiCapabilities: {
    visualize: {
      save: true,
    },
  },
}));

import React from 'react';
import {
  HELLO_WORLD_EMBEDDABLE,
  HelloWorldEmbeddableFactory,
  HelloWorldContainer,
} from '../../../../__test__/index';

import { AddPanelFlyout } from './add_panel_flyout';
import { Container } from 'plugins/embeddable_api/containers';
import { EmbeddableFactoryRegistry } from 'plugins/embeddable_api/embeddables';
// @ts-ignore
import { findTestSubject } from '@elastic/eui/lib/test';
import { mountWithIntl, shallowWithIntl } from 'test_utils/enzyme_helpers';

const onClose = jest.fn();
let container: Container;

function createHelloWorldContainer(input = { id: '123', panels: {} }) {
  const embeddableFactories = new EmbeddableFactoryRegistry();
  embeddableFactories.registerFactory(new HelloWorldEmbeddableFactory());
  return new HelloWorldContainer(input, embeddableFactories);
}

beforeEach(() => {
  container = createHelloWorldContainer();
});

test('matches snapshot', async () => {
  const component = shallowWithIntl(<AddPanelFlyout container={container} onClose={onClose} />);

  expect(component).toMatchSnapshot();
});

test('adds a panel to the container', async done => {
  const component = mountWithIntl(<AddPanelFlyout container={container} onClose={onClose} />);

  expect(Object.values(container.getInput().panels).length).toBe(0);

  const { unsubscribe } = container.getInput$().subscribe(input => {
    expect(input.panels).toBeDefined();
    if (input.panels) {
      expect(Object.values(input.panels).length).toBe(1);
      unsubscribe();
    }
    done();
  });

  findTestSubject(component, 'createNew').simulate('click');
  findTestSubject(component, `createNew-${HELLO_WORLD_EMBEDDABLE}`).simulate('click');
});
