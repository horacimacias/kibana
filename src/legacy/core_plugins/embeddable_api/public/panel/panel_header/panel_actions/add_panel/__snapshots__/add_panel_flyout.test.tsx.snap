// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<EuiFlyout
  closeButtonAriaLabel="Closes this dialog"
  data-test-subj="dashboardAddPanel"
  hideCloseButton={false}
  maxWidth={false}
  onClose={[MockFunction]}
  ownFocus={true}
  size="m"
>
  <EuiFlyoutHeader
    hasBorder={true}
  >
    <EuiTitle
      size="m"
    >
      <h2>
        <FormattedMessage
          defaultMessage="Add panels"
          id="kbn.embeddables.addPanelsTitle"
          values={Object {}}
        />
      </h2>
    </EuiTitle>
  </EuiFlyoutHeader>
  <EuiFlyoutBody>
    <SavedObjectFinder
      noItemsMessage="No matching objects found."
      onChoose={[Function]}
      savedObjectMetaData={Array []}
      showFilter={true}
    />
  </EuiFlyoutBody>
  <EuiFlyoutFooter>
    <EuiFlexGroup
      justifyContent="flexEnd"
    >
      <EuiFlexItem
        grow={true}
      >
        <EuiSuperSelect
          compressed={false}
          data-test-subj="createNew"
          fullWidth={false}
          hasDividers={false}
          isInvalid={false}
          onChange={[Function]}
          options={
            Array [
              Object {
                "inputDisplay": <EuiText>
                  <FormattedMessage
                    defaultMessage="Create new ..."
                    id="kbn.embeddables.addPanel.createNewDefaultOption"
                    values={Object {}}
                  />
                </EuiText>,
                "value": "createNew",
              },
              Object {
                "data-test-subj": "createNew-hello_world",
                "inputDisplay": <EuiText>
                  <FormattedMessage
                    defaultMessage="Create new {factoryName}"
                    id="kbn.embeddables.addPanel.createNew"
                    values={
                      Object {
                        "factoryName": "hello_world",
                      }
                    }
                  />
                </EuiText>,
                "value": "hello_world",
              },
            ]
          }
          valueOfSelected="createNew"
        />
      </EuiFlexItem>
    </EuiFlexGroup>
  </EuiFlyoutFooter>
</EuiFlyout>
`;
